takashi-market-maker.md

```diff
@@ -0,0 +1,170 @@
+
+If this becomes a n-sided problem, speed matters quite a lot -- if you
+wait, someone else will snatch up the trade. Also, with more
+competition, the spread between bid and ask might have to tighten, but
+if there's less competition, the bid and ask spread can widen. You can
+code up an algorithm that adjusts automatically to market volume in
+this case.
+
+> Assume you have a limited amount of capital, which you don't want to
+> expend all at once (since there's volatility in asset prices, which may
+> go up +-10% a day), and there might be a larger amount of sell offers
+> than buy offers, for a particular asset. What happens if a stock goes up
+> too much, or down too much in a given day?
+
+If a stock goes up too much or too little in a given day, as a market
+maker, you'd like to avoid the stock (since volatility decreases money
+earned). You might try to calculate volatility and avoid volatile
+assets.
+
+## System Design
+
+> Design a system that listens to a server that sends requests, records
+> them to a database, and executes or passes on each offer, based on if it
+> is being sold for a 1% discount or if someone is willing to buy at a 1%
+> premium.
+
+First, let's ask about durability guarantees. Since we're a taxable
+financial entity, for tax purposes, we'll need to persist every
+executed order to stable storage, like a database.
+
+We can imagine that we would **receive** a potential offer, do some
+logic to check if we'd like to execute, and if so, send an **execute**
+back to the server. The server will then respond back to us with a
+response that tells us whether or not we were fast enough. If we were,
+we can record the result in the database. The database schema might
+look like this:
+
+```sql
+create table transactions (
+  id INTEGER PRIMARY KEY AUTO INCREMENT,
+  created_at DATETIME NOT NULL DEFAULT NOW(),
+  asset_id INTEGER FOREIGN KEY REFERENCES ASSETS.ID,
+  price MONEY NOT NULL,
+  action ENUM { BUY | SELL } NOT NULL,
+);
+```
+
+The database doesn't need to be particularly fast, but it's beneficial
+to be async (so we don't block the main thread). This would also be
+write-only, so we can disable indexes and optimize for writes. We
+also care a lot about durability, so we could have a cluster of nodes
+taken care of by zookeeper, so writes are replicated.
+
+To execute orders that are properly profitable for us, we also need to
+keep the correct price. Let's assume that there is some other service
+we can make RPC calls to that follows this interface:
+
+```json
+GET /{ticker}
+// returns the current price of the asset, time, and asset.
+{
+    "price": 100.00,
+    "time": 1646358424,
+    "ticker": AAPL
+}
+```
+
+We want to make sure we don't have stale data on the current price set
+by the market, because if we do, we might buy or sell the asset at a
+wrong price, losing us money. So we need to find a way to efficiently
+be able to query and update our prices.
```

Comment:
what is reponsible for the refresh of this data?

Reply:
I'd think of another service hooked up to an event broker (like kafka) for pub-sub, since we want to make sure these machines are only keeping track of the current price. These machines would have an asynchronous process updating from the pub-sub machine every minute or so for the assets they are responsible for buying/selling.